docs/madr/1-reassign-evolve.md

```diff
@@ -0,0 +1,105 @@
+# Reassign ANT to Evolve Implementation Kickoff on the ArNS Portal
+
+- Status: Proposed
+- Approvers: [Ariel], [Dylan], [Atticus], [Phil], [David]
+- Date: [2025-01-30]
+- Authors: [Atticus]
+
+## Context and Problem Statement
+
+Currently to evolve the ANT we use AOS `Eval` - this introduces the problem of
+evaluating what the current version of the code is. If we maintain a variable in
+the state of the process containing a version number, that tells us only what
+that version number is. At any point the user - owner - of the ANT may `Eval`
+new code into the process resulting in changes to the process that may not
+reflect what we expect the version number we assigned to reflect. To overcome
+this we dryrun interactions on APIs and analyze the results to see if the APIs
+perform as expected - this is very inefficient as dry run calls to CU's are
+expensive in networking client side, as well as for the CU.
+
+And so it is proposed to leverage the `Reassign-Name` handler in conjunction
+with `_boot` to fork the process state to a whole new ANT, assign that new ANT
+as the registered ANT to the specified ArNS name in the registry, and using the
+module ID of the process to identify process version and capabilities, and
+solely the module ID.
+
+## Decision Drivers
+
+- Improve the performance of ANT loading client side.
+- Ensure the upgrade process is more robust by using immutable module IDs.
+- Reduce computational load on CU infra.
+- Mitigate risks associated with reassigning ANT state and ARNS records.
+
+## Considered Options
+
+1. **Current Approach (Baseline)**
+   - Multiple dry runs per ANT.
+   - ETH compatibility checked via handler analysis.
+   - Upgrade involves loading Lua code.
+2. **Optimized Approach (Proposed Change)**
+   - Use the module ID to evaluate process version.
+   - Upgrade via spawning a new ANT and reassigning ARNS records.
+
+## Decision Outcome
+
+### Chosen Option: Optimized Approach
+
+The optimized approach reduces computational overhead and improves caching for
+better performance. It also shifts to a more robust upgrade mechanism using
+immutable module IDs rather than version detection heuristics. We can do this
+now that we compile and publish our own WASM binaries, rather than using an AOS
+binary and loading our code into it.
+
+### Positive Consequences
+
+- **Performance Improvement:** Reducing dry runs significantly speeds up ANT
```

Comment:
gateways can choose to support certain ANTs based on their module ID

Reply:
[9ec2e7a](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/9ec2e7a81774fbe2108f2e391238ffa0564d58de)